File: src/hollywood/adapters/rss.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

import feedparser
import httpx
from dateutil import parser as date_parser

from ..config import HollywoodSettings
from ..extractors import extract_text_from_html, strip_html_fragment
from ..models import (
    ArticleBodyRow,
    ArticleEntityRow,
    ArticleRow,
    EntityAliasRow,
    EntityKind,
    EntityRow,
    IngestOptions,
    NormalizedBundle,
    RawPayload,
    SourceDefinition,
    canonicalize_url,
    json_dumps,
    make_stable_id,
    normalize_whitespace,
)
from ..storage import HollywoodStorage
from .base import BaseAdapter


def _entry_value(entry: dict[str, Any], key: str, default: str = "") -> str:
    value = entry.get(key, default)
    if isinstance(value, str):
        return value
    return default


class RssAdapter(BaseAdapter):
    def __init__(self, source: SourceDefinition):
        super().__init__(source)

    def fetch_raw_payloads(
        self, settings: HollywoodSettings, options: IngestOptions
    ) -> list[RawPayload]:
        headers = {"User-Agent": settings.user_agent}
        timeout = settings.request_timeout_seconds
        payloads: list[RawPayload] = []
        remaining = options.limit
        with httpx.Client(headers=headers, timeout=timeout, follow_redirects=True) as client:
            for feed_url in self.source.default_urls:
                response = client.get(feed_url)
                response.raise_for_status()
                parsed = feedparser.parse(response.content)
                entries = list(parsed.entries)
                if options.since is not None:
                    filtered = []
                    for entry in entries:
                        published = _entry_value(entry, "published") or _entry_value(
                            entry, "updated"
                        )
                        if not published:
                            filtered.append(entry)
                            continue
                        published_at = date_parser.parse(published)
                        if published_at >= options.since:
                            filtered.append(entry)
                    entries = filtered
                if remaining is not None:
                    selected_entries = entries[:remaining]
                    remaining -= len(selected_entries)
                else:
                    selected_entries = entries

                selected_urls = [
                    canonicalize_url(_entry_value(entry, "link"))
                    for entry in selected_entries
                    if _entry_value(entry, "link")
                ]
                payloads.append(
                    RawPayload(
                        payload_type="feed_xml",
                        logical_id=feed_url,
                        body=response.content,
                        content_type=response.headers.get("content-type", "application/rss+xml"),
                        source_url=feed_url,
                        metadata={
                            "feed_url": feed_url,
                            "selected_urls": selected_urls,
                        },
                        extension=".xml",
                    )
                )

                if options.full_text:
                    for entry in selected_entries:
                        link = _entry_value(entry, "link")
                        if not link:
                            continue
                        article_response = client.get(link)
                        article_response.raise_for_status()
                        canonical_url = canonicalize_url(str(article_response.url))
                        payloads.append(
                            RawPayload(
                                payload_type="article_html",
                                logical_id=canonical_url,
                                body=article_response.content,
                                content_type=article_response.headers.get(
                                    "content-type", "text/html"
                                ),
                                source_url=link,
                                canonical_url=canonical_url,
                                metadata={
                                    "feed_url": feed_url,
                                    "title": _entry_value(entry, "title"),
                                },
                                extension=".html",
                            )
                        )

                if remaining is not None and remaining <= 0:
                    break
        return payloads

    def normalize_raw_records(
        self,
        settings: HollywoodSettings,
        storage: HollywoodStorage,
        run_id: str,
        raw_records: list[dict[str, object]],
    ) -> NormalizedBundle:
        bundle = NormalizedBundle()
        for record in raw_records:
            payload_type = str(record["payload_type"])
            path = Path(str(record["content_path"]))
            metadata = json.loads(str(record["metadata_json"]))
            if payload_type == "feed_xml":
                parsed = feedparser.parse(path.read_bytes())
                selected_urls = set(metadata.get("selected_urls", []))
                for entry in parsed.entries:
                    link = _entry_value(entry, "link")
                    if not link:
                        continue
                    canonical_url = canonicalize_url(link)
                    if selected_urls and canonical_url not in selected_urls:
                        continue
                    article_id = make_stable_id(self.source.source_id, canonical_url)
                    published_raw = _entry_value(entry, "published") or _entry_value(
                        entry, "updated"
                    )
                    published_at = date_parser.parse(published_raw) if published_raw else None
                    author = (
                        _entry_value(entry, "author") or _entry_value(entry, "dc_creator") or None
                    )
                    categories = [tag.get("term", "") for tag in entry.get("tags", [])]
                    article = ArticleRow(
                        article_id=article_id,
                        source_id=self.source.source_id,
                        canonical_url=canonical_url,
                        url=link,
                        title=normalize_whitespace(_entry_value(entry, "title")),
                        author=author,
                        published_at=published_at,
                        summary=strip_html_fragment(_entry_value(entry, "summary")),
                        feed_guid=_entry_value(entry, "id") or canonical_url,
                        license_class=self.source.license_class.value,
                        run_id=run_id,
                        metadata_json=json_dumps(
                            {"feed_url": metadata.get("feed_url"), "categories": categories}
                        ),
                    )
                    bundle.articles.append(article)

                    description_text = strip_html_fragment(_entry_value(entry, "summary"))
                    if description_text:
                        bundle.article_bodies.append(
                            ArticleBodyRow(
                                body_id=make_stable_id(article_id, "feed_description"),
                                article_id=article_id,
                                source_id=self.source.source_id,
                                body_kind="feed_description",
                                text=description_text,
                                raw_record_id=str(record["raw_record_id"]),
                                content_hash=str(record["content_hash"]),
                                license_class=self.source.license_class.value,
                            )
                        )

                    content_items = entry.get("content", [])
                    if content_items:
                        encoded_text = strip_html_fragment(str(content_items[0].get("value", "")))
                        if encoded_text:
                            bundle.article_bodies.append(
                                ArticleBodyRow(
                                    body_id=make_stable_id(article_id, "feed_content"),
                                    article_id=article_id,
                                    source_id=self.source.source_id,
                                    body_kind="feed_content",
                                    text=encoded_text,
                                    raw_record_id=str(record["raw_record_id"]),
                                    content_hash=str(record["content_hash"]),
                                    license_class=self.source.license_class.value,
                                )
                            )

                    if author:
                        entity_id = make_stable_id(
                            self.source.source_id, EntityKind.PERSON.value, author.lower()
                        )
                        bundle.entities.append(
                            EntityRow(
                                entity_id=entity_id,
                                source_id=self.source.source_id,
                                external_id=None,
                                entity_type=EntityKind.PERSON.value,
                                name=author,
                                canonical_name=author.casefold(),
                                license_class=self.source.license_class.value,
                                metadata_json=json_dumps({"role": "author"}),
                            )
                        )
                        bundle.entity_aliases.append(
                            EntityAliasRow(
                                entity_alias_id=make_stable_id(entity_id, author),
                                entity_id=entity_id,
                                source_id=self.source.source_id,
                                alias=author,
                            )
                        )
                        bundle.article_entities.append(
                            ArticleEntityRow(
                                article_entity_id=make_stable_id(article_id, entity_id, "author"),
                                article_id=article_id,
                                entity_id=entity_id,
                                source_id=self.source.source_id,
                                relation="author",
                            )
                        )
            elif payload_type == "article_html":
                article_url = str(record["canonical_url"] or record["source_url"])
                article_id = make_stable_id(self.source.source_id, canonicalize_url(article_url))
                extracted = extract_text_from_html(
                    path.read_text(encoding="utf-8", errors="replace")
                )
                if extracted:
                    bundle.article_bodies.append(
                        ArticleBodyRow(
                            body_id=make_stable_id(article_id, "page_extract"),
                            article_id=article_id,
                            source_id=self.source.source_id,
                            body_kind="page_extract",
                            text=extracted,
                            raw_record_id=str(record["raw_record_id"]),
                            content_hash=str(record["content_hash"]),
                            license_class=self.source.license_class.value,
                            metadata_json=json_dumps({"source_url": record["source_url"]}),
                        )
                    )
        return bundle

File: api/src/ingest/adapters/rss.ts
import { readFileSync } from "node:fs";
import Parser from "rss-parser";
import { env } from "../../env.js";
import { stripHtmlFragment, extractTextFromHtml } from "../extractors.js";
import {
  canonicalizeUrl,
  emptyBundle,
  extendBundle,
  makeStableId,
  normalizeWhitespace,
} from "../models.js";
import type {
  ArticleContentRow,
  ArticleEntityRow,
  ArticleRow,
  EntityAliasRow,
  EntityRow,
  IngestOptions,
  NormalizedBundle,
  RawPayload,
  SourceDefinition,
} from "../models.js";
import type { Adapter } from "./base.js";
import type { DbRow } from "../../db/index.js";

interface FeedItem {
  title?: string;
  link?: string;
  pubDate?: string;
  isoDate?: string;
  creator?: string;
  content?: string;
  contentSnippet?: string;
  summary?: string;
  guid?: string;
  categories?: string[];
}

function entryPublished(item: FeedItem): string {
  return item.isoDate ?? item.pubDate ?? "";
}

export class RssAdapter implements Adapter {
  private parser = new Parser<object, FeedItem>({
    customFields: { item: [["dc:creator", "creator"]] },
  });

  constructor(public source: SourceDefinition) {}

  async fetchRawPayloads(options: IngestOptions): Promise<RawPayload[]> {
    const headers = { "User-Agent": env.HOLLYWOOD_USER_AGENT };
    const payloads: RawPayload[] = [];
    let remaining = options.limit;

    for (const feedUrl of this.source.defaultUrls) {
      const response = await fetch(feedUrl, { headers, signal: AbortSignal.timeout(env.HOLLYWOOD_REQUEST_TIMEOUT_SECONDS * 1000) });
      if (!response.ok) throw new Error(`Failed to fetch feed ${feedUrl}: ${response.status}`);
      const bodyText = await response.text();
      const feed = await this.parser.parseString(bodyText);
      let entries = feed.items ?? [];

      if (options.since) {
        entries = entries.filter((entry) => {
          const published = entryPublished(entry);
          if (!published) return true;
          const publishedAt = new Date(published);
          return isNaN(publishedAt.getTime()) || publishedAt >= options.since!;
        });
      }

      let selectedEntries = entries;
      if (remaining !== undefined) {
        selectedEntries = entries.slice(0, remaining);
        remaining -= selectedEntries.length;
      }

      const selectedUrls = selectedEntries.filter((e) => e.link).map((e) => canonicalizeUrl(e.link!));

      payloads.push({
        payloadType: "feed_xml",
        logicalId: feedUrl,
        body: Buffer.from(bodyText, "utf-8"),
        contentType: response.headers.get("content-type") ?? "application/rss+xml",
        sourceUrl: feedUrl,
        fetchedAt: new Date(),
        metadata: { feed_url: feedUrl, selected_urls: selectedUrls },
        extension: ".xml",
      });

      if (options.fullText) {
        for (const entry of selectedEntries) {
          if (!entry.link) continue;
          const articleResponse = await fetch(entry.link, { headers, signal: AbortSignal.timeout(env.HOLLYWOOD_REQUEST_TIMEOUT_SECONDS * 1000) });
          if (!articleResponse.ok) continue;
          const articleBody = Buffer.from(await articleResponse.arrayBuffer());
          const canonicalUrl = canonicalizeUrl(articleResponse.url || entry.link);
          payloads.push({
            payloadType: "article_html",
            logicalId: canonicalUrl,
            body: articleBody,
            contentType: articleResponse.headers.get("content-type") ?? "text/html",
            sourceUrl: entry.link,
            canonicalUrl,
            fetchedAt: new Date(),
            metadata: { feed_url: feedUrl, title: entry.title ?? "" },
            extension: ".html",
          });
        }
      }

      if (remaining !== undefined && remaining <= 0) break;
    }

    return payloads;
  }

  async normalizeRawRecords(runId: string, rawRecords: DbRow[]): Promise<NormalizedBundle> {
    const bundle = emptyBundle();
    for (const record of rawRecords) {
      const payloadType = String(record["payload_type"]);
      const path = String(record["content_path"]);
      const metadata = JSON.parse(String(record["metadata_json"] ?? "{}"));

      if (payloadType === "feed_xml") {
        extendBundle(bundle, await this.normalizeFeedXml(runId, record, path, metadata));
      } else if (payloadType === "article_html") {
        extendBundle(bundle, this.normalizeArticleHtml(record, path));
      }
    }
    return bundle;
  }

  private async normalizeFeedXml(runId: string, record: DbRow, path: string, metadata: { feed_url?: string; selected_urls?: string[] }): Promise<NormalizedBundle> {
    const bundle = emptyBundle();
    const xml = readFileSync(path, "utf-8");
    const feed = await this.parser.parseString(xml);
    const selectedUrls = new Set(metadata.selected_urls ?? []);

    for (const entry of feed.items ?? []) {
      if (!entry.link) continue;
      const canonicalUrl = canonicalizeUrl(entry.link);
      if (selectedUrls.size && !selectedUrls.has(canonicalUrl)) continue;

      const articleId = makeStableId(this.source.sourceId, canonicalUrl);
      const publishedRaw = entryPublished(entry);
      const publishedAt = publishedRaw ? new Date(publishedRaw) : undefined;
      const author = entry.creator || undefined;
      const categories = entry.categories ?? [];

      const article: ArticleRow = {
        articleId,
        sourceId: this.source.sourceId,
        canonicalUrl,
        url: entry.link,
        title: normalizeWhitespace(entry.title ?? ""),
        author,
        publishedAt: publishedAt && !isNaN(publishedAt.getTime()) ? publishedAt : undefined,
        summary: stripHtmlFragment(entry.summary ?? entry.contentSnippet),
        feedGuid: entry.guid || canonicalUrl,
        licenseClass: this.source.licenseClass,
        runId,
        metadataJson: JSON.stringify({ feed_url: metadata.feed_url, categories }),
      };
      bundle.articles.push(article);

      const descriptionText = stripHtmlFragment(entry.summary ?? entry.contentSnippet);
      if (descriptionText) {
        const row: ArticleContentRow = {
          contentId: makeStableId(articleId, "feed_description"),
          articleId,
          sourceId: this.source.sourceId,
          contentKind: "feed_description",
          text: descriptionText,
          rawRecordId: String(record["id"]),
          contentHash: String(record["content_hash"]),
          licenseClass: this.source.licenseClass,
          metadataJson: "{}",
        };
        bundle.articleContent.push(row);
      }

      if (entry.content) {
        const encodedText = stripHtmlFragment(entry.content);
        if (encodedText) {
          const row: ArticleContentRow = {
            contentId: makeStableId(articleId, "feed_content"),
            articleId,
            sourceId: this.source.sourceId,
            contentKind: "feed_content",
            text: encodedText,
            rawRecordId: String(record["id"]),
            contentHash: String(record["content_hash"]),
            licenseClass: this.source.licenseClass,
            metadataJson: "{}",
          };
          bundle.articleContent.push(row);
        }
      }

      if (author) {
        const entityId = makeStableId(this.source.sourceId, "person", author.toLowerCase());
        const entityRow: EntityRow = {
          entityId,
          sourceId: this.source.sourceId,
          entityType: "person",
          name: author,
          canonicalName: author.toLowerCase(),
          licenseClass: this.source.licenseClass,
          metadataJson: JSON.stringify({ role: "author" }),
        };
        bundle.entities.push(entityRow);
        const aliasRow: EntityAliasRow = {
          entityAliasId: makeStableId(entityId, author),
          entityId,
          sourceId: this.source.sourceId,
          alias: author,
        };
        bundle.entityAliases.push(aliasRow);
        const articleEntityRow: ArticleEntityRow = {
          articleEntityId: makeStableId(articleId, entityId, "author"),
          articleId,
          entityId,
          sourceId: this.source.sourceId,
          relation: "author",
          metadataJson: "{}",
        };
        bundle.articleEntities.push(articleEntityRow);
      }
    }
    return bundle;
  }

  private normalizeArticleHtml(record: DbRow, path: string): NormalizedBundle {
    const bundle = emptyBundle();
    const articleUrl = String(record["canonical_url"] ?? record["source_url"]);
    const articleId = makeStableId(this.source.sourceId, canonicalizeUrl(articleUrl));
    const html = readFileSync(path, "utf-8");
    const extracted = extractTextFromHtml(html);
    if (extracted) {
      const row: ArticleContentRow = {
        contentId: makeStableId(articleId, "page_extract"),
        articleId,
        sourceId: this.source.sourceId,
        contentKind: "page_extract",
        text: extracted,
        rawRecordId: String(record["id"]),
        contentHash: String(record["content_hash"]),
        licenseClass: this.source.licenseClass,
        metadataJson: JSON.stringify({ source_url: record["source_url"] }),
      };
      bundle.articleContent.push(row);
    }
    return bundle;
  }
}
